File: Autocomplete_Trie.py
"""
Operation: AutocompleteSystem(["i love you", "island","ironman", "i love leetcode"], [5,3,2,2]) 
The system have already tracked down the following sentences and their corresponding times: 
"i love you" : 5 times 
"island" : 3 times 
"ironman" : 2 times 
"i love leetcode" : 2 times 
Now, the user begins another search: 

Operation: input('i') 
Output: ["i love you", "island","i love leetcode"] 
Explanation: 
There are four sentences that have prefix "i". Among them, "ironman" and "i love leetcode" have same hot degree. Since ' ' has ASCII code 32 and 'r' has ASCII code 114, "i love leetcode" should be in front of "ironman". Also we only need to output top 3 hot sentences, so "ironman" will be ignored. 

Operation: input(' ') 
Output: ["i love you","i love leetcode"] 
Explanation: 
There are only two sentences that have prefix "i ". 

Operation: input('a') 
Output: [] 
Explanation: 
There are no sentences that have prefix "i a". 

Operation: input('#') 
Output: [] 
Explanation: 
The user finished the input, the sentence "i a" should be saved as a historical sentence in system. And the following input will be counted as a new search.
"""

##Time Complexity : O(l*n + mlog3) where l is the average length of each sentence and n is the number of given sentences, and m is the average length of the given input string
#Space Complexity : O(l*n) where l is the average length of each sentence and n is the number of given sentences


import heapq
class Pair:
    def __init__(self, sentence, count):
        self.sentence = sentence
        self.count = count
        
        
class TrieNode:
    def __init__(self):
        self.hashmap = {}
        self.children = {}
        
        
class AutocompleteSystem:

    def __init__(self, sentences: List[str], times: List[int]):
        self.root = TrieNode()
        self.input1=""
        for i in range(len(sentences)):
            self.insert(sentences[i], times[i])
         #comparisons based on the frequency and if frequency is same then lexographically compare
        Pair.__lt__ = lambda self, next: self.sentence > next.sentence if self.count == next.count else self.count < next.count
        
    
    def insert(self, sentence, times):
        curr = self.root
        for ch in sentence:
            if ch not in curr.children:
                curr.children[ch] = TrieNode()
            curr = curr.children[ch]
            curr.hashmap[sentence] = curr.hashmap.get(sentence, 0) + times
    
    def search(self, string):
        curr = self.root
        for ch in string:
            if ch not in curr.children:
                return {}
            curr = curr.children[ch]

        return curr.hashmap
    
    def input(self, c: str) -> List[str]:
        if(c=="#"):
            self.insert(self.input1, 1)
            self.input1=""
            return []
        else:
            self.input1+=c
            pq=[]
            sentmap = self.search(self.input1)
            #get top three highest frequency sentences for input
            for sent, count in sentmap.items():
                heapq.heappush(pq, Pair(sent, count))
                if(len(pq)>3):
                    heapq.heappop(pq)
            result=[]
            while(pq):
                pair = heapq.heappop(pq)
                result.insert(0,pair.sentence)    
            return result        

        
obj = AutocompleteSystem(["i love you", "island","ironman", "i love leetcode"], [5,3,2,2])
c='i'
param_1 = obj.input(c)
print(param_1)

File: AutocompleteTrie.java
class AutocompleteSystem {
    class TrieNode{
        TrieNode[] children;
        List<String> li;
        public TrieNode(){
            this.children = new TrieNode[256];//as space need to be considered
            this.li = new ArrayList<>();
        }
    }
    
    private void insert(String sentence)
    {
      TrieNode curr = root;
      for(int i=0;i<sentence.length();i++)
      {
          char c = sentence.charAt(i);
          if(curr.children[c-' '] == null)
          {
              curr.children[c-' '] = new TrieNode();
          }
          curr = curr.children[c-' '];
          curr.li.add(sentence);
      }
    }
    
    private List<String> search(String prefix)
    {
        TrieNode curr = root;
       for(int i=0;i<prefix.length();i++)
       {
           char c = prefix.charAt(i);
           if(curr.children[c-' '] == null)
           {
               return new ArrayList<>();
           }
           curr = curr.children[c-' '];
       }
       return curr.li;
    }

    HashMap<String, Integer> map;
    StringBuilder inputs;
    TrieNode root;
    public AutocompleteSystem(String[] sentences, int[] times) {
        this.map = new HashMap<>();
        this.inputs = new StringBuilder();
        this.root = new TrieNode();

        for(int i=0;i<sentences.length;i++)
        {
            String sentence = sentences[i];
            if(!map.containsKey(sentence)) insert(sentence);
            map.put(sentence, map.getOrDefault(sentence,0)+times[i]);
        }

    }
    
    public List<String> input(char c) {
        if(c == '#')
        {
            String search = inputs.toString();
            if(!map.containsKey(search)) insert(search);
            map.put(search, map.getOrDefault(search,0)+1);
            inputs = new StringBuilder();
            return new ArrayList<>();

        }
        inputs.append(c);
        PriorityQueue<String> pq = new PriorityQueue<>((a,b)->{
            if(map.get(a) == map.get(b))//when freq are equal smaller letter will have precendence
            {
                return b.compareTo(a);// 
                /*String myStr1 = "ir";
String myStr2 = "is";
System.out.println(myStr2.compareTo(myStr1)); s2.comapres1 retruns 1 , s1comapres2 retruns -1
  String myStr1 = "ii";//1st
String myStr2 = "im";//4nd
System.out.println(myStr1.compareTo(myStr2));//i-m = -4*/
            }
            else
            {
                return map.get(a) - map.get(b);//string with lesser freq should be on the top
            }
         });
         String prefix = inputs.toString();
         List<String> startsWith = search(prefix);
         
         for(String str : startsWith)
         {
             if(str.startsWith(prefix))
             {
                 pq.add(str);
             }
             if(pq.size() > 3)
             {
                 pq.poll();
             }
         }
         

        List<String> result = new ArrayList<>();
        while(!pq.isEmpty())
        {
            result.add(0,pq.poll());
        }
        return result;
    }
}

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem obj = new AutocompleteSystem(sentences, times);
 * List<String> param_1 = obj.input(c);
 */